### ISSUE TITLE
passing data that already has missing values 

### ISSUE THREAD

**Issue by csetraynor** [rcamino/imputation-dgm#2]
Hi, I have a use-case where I would need to pass data with missing values to the different methods. I see that because the project is built for estimating the error of simulated missingness, it is not immediately possible to handle such a dataset. Could you please advice me on what would be the best way to change the project so can handle such a straightforward task?
Thanks, thumbs up!

**Comment by csetraynor** [comment]
hi @rcamino just to double-check with you that you know the architecture of the project. If one passes the data substituting the NaN for valid inputs (say 0,1 for binary variables etc) and then changes slightly the generate_mask_for imputation function to something like:
```
def generate_mask_for_2(features, missing_indicator, variable_sizes):
    num_samples = len(features)
    variable_masks = []
    i = 0
    for variable_size in variable_sizes:
        variable_mask = ( torch.reshape( missing_indicator[i], (num_samples, 1 ) ) ).float()
        if variable_size > 1:
            variable_mask = variable_mask.repeat(1, variable_size)
        variable_masks.append(variable_mask)
        i = i+1
    return torch.cat(variable_masks, dim=1)
```
where missing_indicator is a tensor indicator of missing values and changes the options.missing_probability the argument to missing_index, pass as a torch object like:
```
missing_index = torch.load(options.missing_index)
```
would that be sufficient? It seems to work, but I am suspicious, and I would like to know your opinion on this relatively simple hack.

**Comment by rcamino** [comment]
Hi @csetraynor, I am very sorry for the long wait.
My life was a mess for the past months.

I guess you solved it by now, but just in case, here is an example of what I did in a newer project for datasets that have NaNs:

``` python
def compose_with_mask(mask: Tensor, where_one: Optional[Tensor] = None, where_zero: Optional[Tensor] = None,
                      differentiable: bool = False) -> Tensor:
    # use differentiable when NaNs are NOT present in the sources (also required for training)
    if differentiable:
        # in this case both sources need to be present
        assert where_one is not None
        assert where_zero is not None

        return mask * where_one + inverse_mask(mask) * where_zero

    # use non-differentiable when NaNs are present in the sources
    else:
        # the results starts with zeros
        composed = torch.zeros_like(mask)

        # to fill the result where the mask is one
        if where_one is not None:
            # check shape
            assert where_one.shape == mask.shape
            # put zeros in the source where the mask is zero (because we want the values where the mask is one)
            # and add the to the result
            composed += fill_where_mask_is_one(where_one, inverse_mask(mask), 0.0)

        # to fill the result where the mask is zero
        if where_zero is not None:
            # check shape
            assert where_zero.shape == mask.shape
            # put zeros in the source where the mask is one (because we want the values where the mask is zero)
            # and add the to the result
            composed += fill_where_mask_is_one(where_zero, mask, 0.0)

        return composed
```

Some studies fill the missing values with zeros, others with random values, and others with the mean or the mode per feature.
You can interpret any of these alternatives as using an initial imputation.
On the other hand, you could also interpret using zeros as cutting connections like dropout does.
You could also pay special attention in how to generate filling values for the categorical variables that are one-hot-encoded: should you respect or not the one-hot-encoding structure? (i.e picking only one feature at random and setting the rest on zero vs. just setting noise for every feature).

Edit: you can see [here](https://github.com/rcamino/deep-generative-models/blob/master/deep_generative_models/imputation/masks.py) the other methods.

**Comment by csetraynor** [comment]
Hi @rcamino thank you for your feedback no problem these last months have been stressing for everyone. Thank you also for the link to the mask methods that will be very useful.
Best
